### PR TITLE
Add cluster scoped caches

### PR DESCRIPTION
**Pull request by varshaprasad96**
This PR adds Cluster scoped caches to pkg/caches folder. While setting up controller, we can create a cluster scoped cache which stores ClusterName -> Respective cache for that cluster.

The nest step is to modify Rest API endpoint to look at specific clusters. Currently we look only at top level cluster (admin). The watch must be configured to look at `cluster/*` endpoint to get events across all clusters.

### CODE REVIEW

File: pkg/cache/internal/cache_reader.go [hunk 1]
```diff
@@ -179,11 +184,11 @@ func (c *CacheReader) List(_ context.Context, out client.ObjectList, opts ...cli
...
 // String to allow keeping the key format easily in sync with
 // MetaNamespaceKeyFunc.
-func objectKeyToStoreKey(k client.ObjectKey) string {
+func objectKeyToStoreKey(k client.ObjectKey, clusterName string) string {
 	if k.Namespace == "" {
-		return k.Name
+		return k.Name + clusterName
 	}
-	return k.Namespace + "/" + k.Name
+	return k.Namespace + "/" + k.Name + "/" + clusterName
```
Comment (varshaprasad96) [inline pkg/cache/internal/cache_reader.go]: TODO: @varshaprasad96 (Revert this change)

File: pkg/cache/clusterscoped_cache.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	toolscache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// a new global cluster cache to handle cluster scoped resources.
+const globalClusterCache = "_cluster"
```
Comment (varshaprasad96) [inline pkg/cache/clusterscoped_cache.go:33]: We don't need a `globalClusterCache` too eventually. Instead of having a separate global cache, we can iterate over all the cluster caches and return objects from each one of them.

File: examples/kcp/main.go [hunk 3]
```diff
@@ -43,21 +47,47 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 	log.V(1).Info("reconciling chaos pod")
 
-	log.Info(fmt.Sprintf("%+v\n\n%+v\n", ctx, req))
+	fmt.Println("***************************************")
+	fmt.Println(req.ClusterName)
+	fmt.Println("***************************************")
+	// log.Info(fmt.Sprintf("%+v\n\n%+v\n", ctx, req))
 
-	var chaospod api.ChaosPod
+	chaospod := api.ChaosPod{ObjectMeta: metav1.ObjectMeta{ClusterName: req.ClusterName}}
```
Comment (fabianvf) [inline examples/kcp/main.go:55]: this is fine for the prototype but it does feel a little janky, I wonder if there's any safe way we could do something more like:

```
r.Scope(req.ClusterName).Get(...)
```

in the future to scope an individual request rather than requiring the user to partially fill in the metadata before making the request

File: pkg/cache/clusterscoped_cache.go (hunk 2 above)
Comment (fabianvf) [reply pkg/cache/clusterscoped_cache.go:33]: hmm, we'd still need a cache on `*` because multi-cluster requests would go there and could contain information not yet included in any specific cluster's cache, though we could just treat `*` like any other cache since it's not keyed differently. I guess the only other open question here is what we do about the top level admin-cluster scope - we may eventually want to set a cache on `[::1]:6443` in addition to all the `[::1]:6443/clusters/{cluster}` caches

File: pkg/cache/clusterscoped_cache.go [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+		return &multiClusterCache{clusterToCache: caches, Scheme: opts.Scheme, RESTMapper: opts.Mapper, gClusterCache: gCache}, nil
+	}
+}
+
+// multiClusterCache knows how to handle multiple namespaced caches
+// Use this feature when scoping permissions for your
+// operator to a list of namespaces instead of watching every namespace
+// in the cluster.
+type multiClusterCache struct {
+	clusterToCache map[string]Cache
```
Comment (fabianvf) [inline pkg/cache/clusterscoped_cache.go:77]: so I guess what I mean is `*` would still fit this data structure

File: pkg/cache/clusterscoped_cache.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	Scheme         *runtime.Scheme
+	RESTMapper     apimeta.RESTMapper
+	gClusterCache  Cache // Point to "*"
```
Comment (fabianvf) [inline pkg/cache/clusterscoped_cache.go:80]: and this could become `kcpClusterCache` or `adminClusterCache` and just point to the actual scope in the unmodified rest.Config that we receive. Probably would need to talk all that over with the kcp folks and see if that's even something we'd want to allow from c-r by default.

File: pkg/cache/clusterscoped_cache.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+		opts, err := defaultOpts(config, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		caches := map[string]Cache{}
+
+		// create aglobal cache for * scope
+		globalConfig := *config
+		globalConfig.Host = globalConfig.Host + "/clusters/*"
```
Comment (fabianvf) [inline pkg/cache/clusterscoped_cache.go:52]:
```suggestion
		globalConfig := *config
		//TODO: move this rest.Config.Host hacking to a helper method
		globalConfig.Host = globalConfig.Host + "/clusters/*"
```

File: pkg/cache/clusterscoped_cache.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+		gCache, err := New(&globalConfig, opts)
+		if err != nil {
+			return nil, fmt.Errorf("error creating global cache %v", err)
+		}
+
+		for _, cs := range clusterNames {
+			scopedConfig := *config
+			scopedConfig.Host = config.Host + "/clusters/" + cs
```
Comment (fabianvf) [inline pkg/cache/clusterscoped_cache.go:60]:
```suggestion
			//TODO: move this rest.Config.Host hacking to a helper method
			scopedConfig.Host = config.Host + "/clusters/" + cs
```

File: pkg/cache/clusterscoped_cache.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+	if (clusterName) == "*" {
+		globalInformer, err := c.gClusterCache.GetInformer(ctx, obj)
+		if err != nil {
+			return nil, err
+		}
+		informers[globalClusterCache] = globalInformer
+	}
+
+	for cs, cache := range c.clusterToCache {
+		informer, err := cache.GetInformer(ctx, obj)
```
Comment (fabianvf) [inline pkg/cache/clusterscoped_cache.go:104]: are we properly scoping the `rest.Config.Host` here?

File: pkg/cache/clusterscoped_cache.go (hunk 2 above)
Comment (varshaprasad96) [reply pkg/cache/clusterscoped_cache.go:33]: The `*` cluster scope is only to list and watch, so we can iterate over all the other caches to get that info. What did you mean by, `information not yet included in any specific cluster's cache` ?
